devito/dimension.py

```diff
@@ -322,6 +322,10 @@ def symbolic_start(self):
     def symbolic_end(self):
         return self._interval.right
 
+    @property
+    def symbolic_size(self):
```

Comment:
Eek.  I think I may have conflated `size` vs `extent`.  Will go back and double check what I've done.  But I think `Dimension.symbolic_extent` should be fixed, right?

Reply:
In fact, I think we should drop it. The extent is a concept that should belong to an ``Iteration``, not to a ``Dimension``, which only provides a start (min) and end (max) points. Whether the interval is bounded or unbounded or closed or whatever, should be responsibility of whoever is using the ``Dimension``.

I'm using it here:
```
 devito/dle/backends/advanced.py:                
    binnersize = i.dim.symbolic_extent + (i.offsets[1] - i.offsets[0])
```

but this one could/should be switched to ``i.symbolic_extent``. 

I see that in ``autotuner.py``, where you're using this property, we do have the ``iterations``. So the information is also available there.

On an aside, we have the ``symbolic_*`` methods in ``dimension.py`` (e.g., ``symbolic_start``, but we call them ``*_symbolic`` in ``nodes.py`` (e.g., ``start_symbolic``). We should fix this

So if changing all this looks too much work, feel free to open an issue

Or if you have objections or other viewpoints, I'm all ears!

Reply:
Am I correct in thinking that `size` is only relevant when storage (e.g. a `Function`) is declared with a given `Dimension`?

So `SubDimension.symbolic_size` should never be used, right now? - until `Function`s declared on `SubDimension`s are supported at least ;-)

Reply:
so, ``symbolic_size`` in the superclass is necessary because it provides the *name* of the size variable, not the actual value

Reply:
But for a SubDimension that name should never be used, anymore.

Reply:
Perhaps a SubDimension should return the parent's `symbolic_size`?

Reply:
yeah, that might be an idea. It might even be the case for all ``DerivedDimension``.

Reply:
I feel like that change (returning the parent's symbolic_size) is probably unnecessary right now?  Nothing fails because of this, today.

Further, I feel like it might even be a step backwards: in future, it will very likely be necessary to declare functions which are only defined on a `SubDimension` (e.g. for keeping full time history for boundaries, or more sophisticated PMLs).  In this case, the size will need to differ from the parent.

Reply:
> Further, I feel like it might even be a step backwards: in future, it will very likely be necessary to declare functions which are only defined on a SubDimension (e.g. for keeping full time history for boundaries, or more sophisticated PMLs). In this case, the size will need to differ from the parent.

OK, this indeed makes sense!

Reply:
Though, on a second thought, perhaps we should just return a name, or perhaps the parent's ``symbolic_size`` minus whatever the ``SubDimension`` doesn't span. Otherwise, the way it is now looks more of an extent.

Reply:
That seems like the most sensible thing.  The storage sizes (used in the casts at the top of the kernel) shouldn't depend on the `x_m` and `x_M`.